### ISSUE TITLE
replace function doesn't fully support all replacement concatenation

### ISSUE THREAD

**Issue by BrandonALXEllisSS** [hashicorp/terraform#28883]
### Terraform Version

```
15.0-beta
```

### Terraform Configuration Files

```
replace("adf", "/(a)/", "$1s")
replace("adf", "/(a)/", "$1 s")
replace("adf", "/(a)/", "$1-s")
```

### Expected Behavior
```
> asdf
> a sdf
> a-sdf
```

### Actual Behavior
```
> df
> a sdf
> a-sdf
```

### Steps to Reproduce

1. terraform console
2. `replace("adf", "/(a)/", "$1s")`

### Additional Context

Since the replace function allows both indexes and numbers, there is no way to tell the difference between named capture groups, indexed capture groups, and bad names for capture groups. Hence the behavior shown above

It seems to delimit on characters that do not make up capture-group name characters

Ideally there should be some sort of delimiter for replacement subgroups so that direct string concatenation is possible. (i.e., something like $(index) or $(name) instead of $index or $name

**Comment by jbardin** [comment]
Hi @BrandonALXEllisSS,

The common syntax for delimiting named matching groups is to use `${...}`. The trick here is that because regular expressions are a language themselves, and you need to embed that string within HCL, you have to escape the `${` pattern so that it is passed through unchanged, and HCL escapes `$` by doubling them. In the case of your example, you can use:

    replace("adf", "/(a)/", "$${1}s")


Duplicate of #27937

Thanks!

**Comment by BrandonALXEllisSS** [comment]
Where is this in the documentation???

Going off of the reference pages alone, there's no way to deduce that replace is using the "The common syntax for delimiting named matching groups". 

The [documentation page on "replace"](https://www.terraform.io/docs/language/functions/replace.html) makes it sounds like this is a special one-off exception and **never even mentions curly brackets anywhere**.  
> the replacement string can incorporate captured strings from the input by using an $n sequence, where n is the index or name of a capture group

On the other hand, the documentation page on strings doesn't mention anything about string interpolation without curly brackets like how the replace function uses it (the "$n" syntax).

It's really hard to believe that these use the same kind of string interpolation and it seems like the documentation of the "replace" function should be updated to make this more apparent.

**Comment by jbardin** [comment]
Thanks @BrandonALXEllisSS!

The `replace` function documentation does reference that it uses the same syntax as the `regex` function, which more thoroughly covers common expressions. String interpolation itself is a function of the HCL language used by terraform, and therefore all strings are treated equally in that respect, so any possible interpolation before passing the value to the given function must happen first.

The interpolation of matching group indexes by the `replace` function is not a property of HCL, but a feature of the regular expression implementation.  We do not provide comprehensive documentation on regular expression syntax, and some prior knowledge of regular expressions is expected. However name group replacement syntax is less commonly shown as it's not directly part of the regular expression, so it would probably help to document that on the `replace` page for future readers. 

To add one more example here as I'm re-opening this for documentation, the equivalent named group syntax would look like so:

    replace("adf", "/(?P<foo>a)/", "$${foo}s")

**Comment by apparentlymart** [comment]
I must admit that I (who wrote much of the current `replace` documentation text) actually had no idea that this named replacement syntax (and the brace-based variant of it) was possible, and so that's the reason why it's not documented!

The design of `replace` is kinda unfortunate in a number of ways, due to it being a long-standing function from back before our overall language revamp in v0.12. We typically try to avoid situations where there's HCL-like syntax embedded in strings that you'd then need to escape, but this one's old and so too late now to change it. We certainly can document it better, though.

If we had the luxury of designing this over again I'd consider using a custom decoding rule for that third argument so that we can integrate the regular expression values directly into our usual language syntax, something like `${regex.foo}`, but we only got the ability to do that sort of thing relatively recently and so we're stuck with the current design for now.
